packages/compiler/core/parser.ts

```diff
@@ -541,6 +538,23 @@ export function parse(code: string | SourceFile, options: ParseOptions = {}): Ca
     });
   }
 
+  function parseTemplateParameterList(): TemplateParameterDeclarationNode[] {
+    const list = parseOptionalList(ListKind.TemplateParameters, parseTemplateParameter);
+    let setDefault = false;
```

Comment:
This actually seems fine to me to do as a separate pass like this. I suppose you could also pass a stateful wrapper around parseTemplateParameter to keep one pass.

something like
```js
let setDefault = false;
const parseOptionalList(ListKind.TemplateParameters, () => {
  const item = parseTemplateParameter();
  if (!item.default && setDefault) {
    error({ code: "default-required", target: item });
  }

  if (item.default) {
    setDefault = true;
  }
  return item;
});
```

Not sure if that's better or not, though.

Reply:
Don't know why passing a callback like that didn't occur to me but that's a great solution thank you. That said, if you're happy with two-pass it's a bit simpler so happy to leave as is!